File: binding/include/libqversionnumber.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qversionnumber.html
pub const qversionnumber = struct {

    /// New constructs a new QVersionNumber object.
    pub fn New() QtC.QVersionNumber {
        return qtc.QVersionNumber_new();
    }


    /// New2 constructs a new QVersionNumber object.
    pub fn New2(args: QSpan<const int>) QtC.QVersionNumber {
        return qtc.QVersionNumber_new2(args);
    }


    /// New3 constructs a new QVersionNumber object.
    pub fn New3(maj: i32) QtC.QVersionNumber {
        return qtc.QVersionNumber_new3(maj);
    }


    /// New4 constructs a new QVersionNumber object.
    pub fn New4(maj: i32, min: i32) QtC.QVersionNumber {
        return qtc.QVersionNumber_new4(maj, min);
    }


    /// New5 constructs a new QVersionNumber object.
    pub fn New5(maj: i32, min: i32, mic: i32) QtC.QVersionNumber {
        return qtc.QVersionNumber_new5(maj, min, mic);
    }


    pub fn IsNull(self: ?*anyopaque, ) bool {
        return qtc.QVersionNumber_IsNull(@ptrCast(self));
    }

    pub fn IsNormalized(self: ?*anyopaque, ) bool {
        return qtc.QVersionNumber_IsNormalized(@ptrCast(self));
    }

    pub fn MajorVersion(self: ?*anyopaque, ) i32 {
        return qtc.QVersionNumber_MajorVersion(@ptrCast(self));
    }

    pub fn MinorVersion(self: ?*anyopaque, ) i32 {
        return qtc.QVersionNumber_MinorVersion(@ptrCast(self));
    }

    pub fn MicroVersion(self: ?*anyopaque, ) i32 {
        return qtc.QVersionNumber_MicroVersion(@ptrCast(self));
    }

    pub fn Normalized(self: ?*anyopaque, ) QtC.QVersionNumber {
        return qtc.QVersionNumber_Normalized(@ptrCast(self));
    }

    pub fn Segments(self: ?*anyopaque, ) []const u8 {
        return qtc.QVersionNumber_Segments(@ptrCast(self));
    }

    pub fn SegmentAt(self: ?*anyopaque, index: i64) i32 {
        return qtc.QVersionNumber_SegmentAt(@ptrCast(self), index);
    }

    pub fn SegmentCount(self: ?*anyopaque, ) i64 {
        return qtc.QVersionNumber_SegmentCount(@ptrCast(self));
    }

    pub fn IsPrefixOf(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QVersionNumber_IsPrefixOf(@ptrCast(self), @ptrCast(other));
    }

    pub fn Compare(v1: ?*anyopaque, v2: ?*anyopaque) i32 {
        return qtc.QVersionNumber_Compare(@ptrCast(v1), @ptrCast(v2));
    }

    pub fn CommonPrefix(v1: ?*anyopaque, v2: ?*anyopaque) QtC.QVersionNumber {
        return qtc.QVersionNumber_CommonPrefix(@ptrCast(v1), @ptrCast(v2));
    }

    pub fn ToString(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QVersionNumber_ToString(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qversionnumber.ToString: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QVersionNumber_Delete(@ptrCast(self));
    }
};

File: binding/src/libqversionnumber.cpp
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QVersionNumber>
#include <qversionnumber.h>
#include "libqversionnumber.h"
#include "libqversionnumber.hxx"

QVersionNumber* QVersionNumber_new() {
	 return new QVersionNumber();
}

QVersionNumber* QVersionNumber_new2(libqt_list args) {
	 return new QVersionNumber(QSpan<const int>());
}

QVersionNumber* QVersionNumber_new3(int maj) {
	 return new QVersionNumber(maj);
}

QVersionNumber* QVersionNumber_new4(int maj, int min) {
	 return new QVersionNumber(maj, min);
}

QVersionNumber* QVersionNumber_new5(int maj, int min, int mic) {
	 return new QVersionNumber(maj, min, mic);
}

bool QVersionNumber_IsNull(const QVersionNumber* self) {
	return self->isNull();
}

bool QVersionNumber_IsNormalized(const QVersionNumber* self) {
	return self->isNormalized();
}

int QVersionNumber_MajorVersion(const QVersionNumber* self) {
	return self->majorVersion();
}

int QVersionNumber_MinorVersion(const QVersionNumber* self) {
	return self->minorVersion();
}

int QVersionNumber_MicroVersion(const QVersionNumber* self) {
	return self->microVersion();
}

QVersionNumber* QVersionNumber_Normalized(const QVersionNumber* self) {
	return new QVersionNumber(self->normalized());
}

libqt_list QVersionNumber_Segments(const QVersionNumber* self) {
	auto _ret = self->segments();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

int QVersionNumber_SegmentAt(const QVersionNumber* self, ptrdiff_t index) {
	return self->segmentAt(index);
}

ptrdiff_t QVersionNumber_SegmentCount(const QVersionNumber* self) {
	return self->segmentCount();
}

bool QVersionNumber_IsPrefixOf(const QVersionNumber* self, const QVersionNumber* other) {
	return self->isPrefixOf(*other);
}

int QVersionNumber_Compare(const QVersionNumber* v1, const QVersionNumber* v2) {
	return QVersionNumber::compare(*v1, *v2);
}

QVersionNumber* QVersionNumber_CommonPrefix(const QVersionNumber* v1, const QVersionNumber* v2) {
	return new QVersionNumber(QVersionNumber::commonPrefix(*v1, *v2));
}

libqt_string QVersionNumber_ToString(const QVersionNumber* self) {
	QString _ret = self->toString();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QVersionNumber_Delete(QVersionNumber* self) {
    delete self;
}

